src/DynamoUtilities/SmartObservableCollection.cs

```diff
@@ -24,15 +27,22 @@ public class SmartObservableCollection<T> : ObservableCollection<T>
         /// </summary>
         internal event NotifyCollectionChangedEventHandler CollectionChangedDuringDeferredReset;
 
-        public SmartObservableCollection() : base()
+        public SmartObservableCollection() : this(null)
         {}
 
-        public SmartObservableCollection(List<T> list)
-            : base(list)
-        {}
-
-        public SmartObservableCollection(IEnumerable<T> collection) : base(collection)
+        public SmartObservableCollection(List<T> list) : this(list as IEnumerable<T>) 
         {}
+  
+        public SmartObservableCollection(IEnumerable<T> collection) : base()
+        {
+            try
+            {
+                //Replace internal IList (from Collection<T>.items) with a thread safe IList(ThreadSafeList)
+                var itemsField = typeof(Collection<T>).GetField("items", BindingFlags.NonPublic | BindingFlags.Instance);
+                itemsField?.SetValue(this, new ThreadSafeList<T>(items: collection));
+            }
+            catch { }
```

Comment:
If this seems like a bit of a hack...that's because it is (there is no protected or public way to change the inner field `items`)

Pretty safe though...All I am doing is replacing the internal (private) IList `items` that Collection<T> is using under the hood. The `items` field is only accessed as an IList...so passing another implementation of IList will work ok
Here is the source code for Collection<T> https://referencesource.microsoft.com/#mscorlib/system/collections/objectmodel/collection.cs,281923b8611114ec

Reply:
The only risk I see...is if Microsoft decides to change their internal workings and maybe rename or remove the private field `Collection<T>.items`

In that case we could leave the SmartObservableCollection<T> to be thread-unsafe...

Reply:
The alternative would be to loose compatibility with the ObservableCollection class (SmartObservableCollection would not be derived from ObservableCollection anymore)